R/list-simplify.R

```diff
@@ -0,0 +1,77 @@
+#' Simplify a list to an atomic or S3 vector
+#'
+#' @details
+#' Simplification maintains a one-to-one correspondence between the input
+#' and output, implying that each element of `x` must contain a vector of
+#' length 1. If you don't want to maintain this correspondence, then you
+#' probably want either [list_c()] or [list_flatten()].
+#'
+#' @param x A list.
+#' @param strict What should happen if simplification fails? If `TRUE`,
+#'   will error. If `FALSE`, will return `x` unchanged.
+#' @param ptype An optional prototype to ensure that the output type is always
+#'   the same.
+#' @returns A vector the same length as `x`.
+#' @export
+#' @examples
+#' list_simplify(list(1, 2, 3))
+#'
+#' try(list_simplify(list(1, 2, "x")))
+#' try(list_simplify(list(1, 2, 1:3)))
+list_simplify <- function(x, strict = TRUE, ptype = NULL) {
+  simplify_impl(x, strict = strict, ptype = ptype)
+}
+
+# Wrapper used by purrr functions that do automatic simplification
+list_simplify_internal <- function(x,
+                                   simplify = NA,
+                                   ptype = NULL,
+                                   error_call = caller_env()
+  ) {
+  if (length(simplify) > 1 || !is.logical(simplify)) {
+    cli::cli_abort("{.arg simplify} must be `TRUE`, `FALSE`, or `NA`.", arg = "simplify")
+  }
+  if (!is.null(ptype) && isFALSE(simplify)) {
+    cli::cli_abort("Can't specify {.arg ptype} when `simplify = FALSE`.")
+  }
+
+  if (isFALSE(simplify)) {
+    return(x)
+  }
+
+  simplify_impl(
+    x,
+    strict = !is.na(simplify),
+    ptype = ptype,
+    error_call = error_call
+  )
+}
+
+simplify_impl <- function(x,
+                          strict = TRUE,
+                          ptype = NULL,
+                          error_call = caller_env()) {
+  vec_check_list(x, call = error_call)
+
+  can_simplify <- every(x, vec_is, size = 1)
```

Comment:
Even if you want to support scalar types, it might be worth using the approach above and wrapping it in `tryCatch()` that returns `FALSE` on error, as it will still probably be an order of magnitude or more faster than `vec_is()`